File: 230_kth_smallest.ts
class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}

function kthSmallest(root: TreeNode | null, k: number): number {
    let count = 0;
    let val = 0;
    function dfs(node: TreeNode | null) {
        if (!node) return;
        dfs(node.left);

        if (++count === k) {
            val = node.val;
            return;
        }

        dfs(node.right);
    }

    dfs(root);
    return val;
}

File: 17thJune2024/230_kth_smallest.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left, *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
};

class Solution {
  int ans;
  int k;
  int i = 0;

public:
  void util(TreeNode *root) {
    if (!root)
      return;

    util(root->left);
    if (i == k) {
      ans = root->val;
    }
    i++;
    util(root->right);
  }

  int kthSmallest(TreeNode *root, int k) {
    this->k = k - 1;
    util(root);
    return this->ans;
  }
};
